app/services/notifications/create_from_model_service.rb

```diff
@@ -292,7 +292,13 @@ def add_receiver(receivers, collection, reason)
   end
 
   def remove_self_recipient(receivers)
-    receivers.delete(user_with_fallback.id)
+    if receivers.key?(user_with_fallback.id)
+      self_reasons = receivers[user_with_fallback.id]
+      self_reasons.delete_if { |item| item != :mentioned }
+      if self_reasons.empty?
+        receivers.delete(user_with_fallback.id)
+      end
+    end
```

Comment:
Since this now becomes more complicated it might be a good idea to extract this into the strategies. That way, it could be used directly in `add_receiver`. That way, each strategy could have an allowlist of reasons self mentioning is accepted. It would be empty for all strategies but would allow self mentioning for work packages. 

Granted, this is not a necessity at this point in time since the code is still acceptable the way it is.

Reply:
I tried to refactor it by using your idea here. Unfortunately, I found that the readability is not improved as much as I would have hoped for. Maybe we can postpone this to a later point in time where we actually know more requirements about self-mentioning.